### PR TITLE
Inline-Editing: allow client to provide custom components

### LINKED ISSUE #29: Inline-Editing: Use a dataComponent for editing cells?
**Problem**
When using the inline editing features, all cells default to a [`<textarea>`](https://github.com/muonw/muonw-powertable/blob/ce50210afc9aead43bb28a9493a8a9e3b100789a/app/src/lib/components/PowerTable.svelte#L1266).

Currently there is a `dataComponent` option for rendering a cell, but not editing a cell.

**Motivation / Goals**
I would like to be able to customize the experience of editing certain cells to allow for different inputs, such as a [selection menu](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select).

**Discussion**
Would this functionality be reasonable to add, or is there a way to accomplish this with the current feature set?
If there is not a way to currently accomplish this, I am happy to take a stab at implementing it.

### PR DESCRIPTION
**Pull request by mavcook**
I took a stab at resolving #29, which allows users to use a custom component when editing a cell in the table. My approach was to create an object for `edit` related configuration in the `Instructs` interface. This gives a user almost full control over how they want editing to work, and leaves room for more capabilities and tuning in the future.

Check the updated `example7/` code for full context, but below gives a quick idea of the end result.
```typescript
let ptInstructs: Instructs[] = [
    {
        key: 'first_name',
    },
    {
        title: 'Favorite Color',
        key: 'favoriteColor', 
        edit: {
            as: <ComponentType<SvelteComponent>>MyDropdownComponent,
            props: {selectValues: ['Red', 'Blue', 'Green']}
        },
    },
]
```
We add a custom component for the `favoriteColor` column that is a dropdown menu
image

## Changes
- `EditBlock` component encapsulates the default textarea edit behavior
- Submitting edits now occurs through a [Svelte Component event](https://learn.svelte.dev/tutorial/component-events)

### Client/User Breakage
- B1. Submitted edits no longer dispatch a `rowClicked` event
- ~~B2. Clients using `dataComponent`s will get warning `<component> was created with unknown prop 'instructs'` (this can be easily reverted)~~

## Notes
N0. This is a larger change than initially anticipated, and as such it introduces slightly different behavior. If the change are too much, or not the right approach, I am happy to rework anything.

N1. ~~Adding the `Cell` interface made documentation and user adoption of the edit component a bit easier, but maybe isn't the right naming or approach. I didn't see any way to inhertance or templating in Svelte~~

----
fixes #29

### CODE REVIEW

File: app/src/lib/components/PowerTable.svelte [hunk 1]
```diff
@@ -21,6 +21,18 @@ export interface Instructs {
     dataComponent?: ComponentType<SvelteComponent>,
 }
 
+export interface Cell {
```
Comment (muonw-public) [inline app/src/lib/components/PowerTable.svelte]: This interface makes the code cleaner but can lower the performance by making a second copy of instructs. I'll elaborate on this later in this file.
This is generally the right way to do things but probably would need a rewrite of other parts.

File: app/src/lib/components/PowerTable.svelte [hunk 2]
```diff
@@ -21,6 +21,18 @@ export interface Instructs {
...
+    value: string,
+    rowIndex: number,
+    rowId: number,
+    instructs: Instructs,
+}
+
+export const EDIT_CONSTS = {
```
Comment (muonw-public) [inline app/src/lib/components/PowerTable.svelte]: This also is a good practice, but at this point we are not optimizing for backward compatibility and can afford to have these values hard coded for simplicity (as they are unlikely to change before a breaking change).

File: app/src/lib/components/PowerTable.svelte [hunk 3]
```diff
@@ -721,31 +733,26 @@ function updatePageSize() {
...
-function manualRowEdit(e: Event, index: number) {
-    let textareaEls = (<HTMLInputElement>e.target).closest('tr')?.querySelectorAll<HTMLInputElement>('textarea[data-name=edit-textarea]');
-    textareaEls?.forEach(textarea => {
-        textarea.style.height = textarea.scrollHeight + 'px';
+function handleSubmittedEdits(event: CustomEvent){
+    let submittedCell: Cell = event.detail.cell
+    let row = data[pageData[submittedCell.rowIndex][dataIdKey]];
+    // find all cells in the row and assign their values back to the pageData
+    let cells = (<HTMLInputElement>event.detail.domEvent.target).closest('tr')?.querySelectorAll(`[data-name=${EDIT_CONSTS.DATA_ATTR_NAME}]`);
+    cells!.forEach(cell => {
```
Comment (muonw-public) [inline app/src/lib/components/PowerTable.svelte]: There should be some sort of error handling in case user doesn't set the `data-name` correctly and therefore element cannot be found.

File: app/src/lib/components/PowerTable.svelte [hunk 4]
```diff
@@ -1171,6 +1178,16 @@ function highlightMatches(pageContent: Data[], matches: HighlightMatches, Instru
     return pageContent;
 }
 
+function composeCell(record: Data, index: number, instructs: Instructs): Cell {
```
Comment (muonw-public) [inline app/src/lib/components/PowerTable.svelte]: This also makes the code inside this file cleaner but is not ideal for performance.

File: app/src/lib/components/PowerTable.svelte [hunk 5]
```diff
@@ -1260,20 +1277,19 @@ onMount(() => {
                                     {#each formattedPageData as record, index}
                                         <tr data-index={index} data-id={record[dataIdKey]} on:click={(e)=>rowClicked(e, index)} on:dblclick={(e)=>rowDblClicked(e, index)}>
                                             {#each instructs as instruct}
+                                                {@const cell = composeCell(record, index, instruct)}
```
Comment (muonw-public) [inline app/src/lib/components/PowerTable.svelte]: This line will run for each cell. For the sample dataset used in Example 7 (which is not that large) this runs 5,970 times in each render. It's not that bad, but it's not ideal since most of the instruct object that it copies is not even useful for editing purposes.

File: app/src/lib/styles/power-table-mascara.scss [hunk 6]
```diff
@@ -64,7 +64,7 @@ div[data-component=powertable]{
     div[data-name=edit-block]{
         label{
             width: 100%;
-            [data-name=edit-textarea]{
+            textarea{
```
Comment (muonw-public) [inline app/src/lib/styles/power-table-mascara.scss]: These are also good to have by default (in addition to textarea):
input:not([type]),
input[type=date],
input[type=datetime-local],
input[type=email],
input[type=month],
input[type=number],
input[type=password],
input[type=tel],
input[type=text],
input[type=time],
input[type=url],
input[type=week],
select

File: app/src/routes/examples/example7/MyComponent.svelte [hunk 7]
```diff
@@ -1,24 +1,32 @@
...
+
+	function submitEdits(e: Event) {
+		dispatch(EDIT_CONSTS.SUBMISSION_EVENT, {
+			cell: cell,
+			domEvent: e
+		});
+	}
 </script>
 
 <label>
```
Comment (muonw-public) [inline app/src/routes/examples/example7/MyComponent.svelte]: To be consistent with the style of the default edit component, this needs to be warped in `<div data-name="edit-block">` and have 'data-name="edit-submit"` the submit button.

File: app/src/routes/examples/example7/MyComponent.svelte [hunk 8]
```diff
@@ -1,24 +1,32 @@
...
 	<span>
-		<span>{ptInstructs?.title}</span>
+		<span>{cell.instructs?.title}</span>
 	</span>
 
-	<select bind:value>
-		{#each selectValues as question}
-			<option value={question}>
-				{question}
+	<select bind:value={cell.value} data-key={cell.instructs.key} data-name={EDIT_CONSTS.DATA_ATTR_NAME} >
```
Comment (muonw-public) [inline app/src/routes/examples/example7/MyComponent.svelte]: The problem with `bind:value` is that it makes the element too reactive for our use-case. If you have more than one row in edit mode (check-marked) and submit your edit for one row the value of the select element in all the other rows will reset to their default. This makes it impossible to change the values of multiple rows and then submit those changes one by one.

File: app/src/routes/examples/example7/+page.svelte [hunk 9]
```diff
@@ -1,22 +1,35 @@
...
+        edit: {
+            as: <ComponentType<SvelteComponent>>MyComponent,
+            props: {selectValues: uniqueDepartments}
+        },
+    },
+    {key: 'job'}
+];
+
+// default a company field to HTML to make capabilities more clear
+data[0]['company'] = `<li>${data[0]['company']}</li><li>html supported</li>`
```
Comment (muonw-public) [inline app/src/routes/examples/example7/+page.svelte]: This might not be the right example page for this demonstration. But more importantly, in many cases it's better not to mix HTML tags with the underlying data if all we want is a better styling. HTML can interfere with text processing (e.g. search and filter). If you change a cell value from `John` to `<li>John</li>` and then filter/search for `li` that cell will show up which is usually not expected.

A proper way to do this for your example is to add a break line between `${data[0]['company']}` and `html supported` and then use a `customParse` to wrap each line in `<li>` tags. This way the HTML tag would not mix with the data.

**Comment by muonw-public** [comment]
Thank you! It's a great implementation. Since you have two commits that somewhat overlap, it might be easier to see the detailed suggested changes here (some of them are just different variable names or styling

**Comment by muonw-public** [comment]
Thank you! It's a great implementation. Since you have two commits that somewhat overlap, it might be easier to see the detailed suggestions here in this temporary branch (some are just for suggesting an alternative name and a few are changes to the parts that where not part of your commits):

https://github.com/muonw/muonw-powertable/commit/724ef81061336a7be93164a99cff54301f5ba54d

The only major change is the removal of the `Cell` interface even though it makes the code cleaner (you will find a comment on the reason for this).

File: app/src/lib/components/PowerTable.svelte (hunk 2 above)
Comment (mavcook) [reply app/src/lib/components/PowerTable.svelte]: I think the greater benefit here is for a client of the library, where you don't want to "maintain" a string yourself. Specifically for the `SUBMISSION_EVENT`. Say I make 4 custom components, I would probably then want to define my own constant for this message name I need to use in my `dispatch()` call. So I think it is much nicer to import from Powertable than define my own.
Would you be ok with just having the `SUBMISSION_EVENT` constant available for users?

File: app/src/routes/examples/example7/MyComponent.svelte (hunk 7 above)
Comment (mavcook) [reply app/src/routes/examples/example7/MyComponent.svelte]: My thinking here was that this example intentionally only implements the minimum markup needed for the functionality to work, not the styling. But I see why it would be good to have the consistency (and also get the CSS rules back).

Would it be worth doing two example components? The fixed version of above, and another component with some custom styling?

File: app/src/routes/examples/example7/+page.svelte (hunk 9 above)
Comment (mavcook) [reply app/src/routes/examples/example7/+page.svelte]: > This might not be the right example page for this demonstration. 

Agreed. I was trying to keep the same functionality of the example before (with the addition of making it clear when viewing the page that the Company column allowed HTML). But seems appropritate to remove since `parseAs: 'html` is used in other examples.

> A proper way to do this for your example...

Awesome, I appreciate you taking the time to teach me a bit here and provide reasoning!

**Comment by mavcook** [comment]
Nice, thank you for the review and changes! It is cool learning more about javascript and Svelte performance.

It looks like you made a lot of the changes in your suggestions commit, so I resolved the related conversations. I will pull your commit into my branch, and I will work on the remaining changes.

File: app/src/routes/examples/example7/MyComponent.svelte (hunk 8 above)
Comment (mavcook) [reply app/src/routes/examples/example7/MyComponent.svelte]: Ah, nice catch, thank you!

File: app/src/lib/components/PowerTable.svelte (hunk 3 above)
Comment (mavcook) [reply app/src/lib/components/PowerTable.svelte]: Fixed in bd5ab850dff1f3dbf6ac080c25fcdac5fccd6008

File: app/src/lib/components/PowerTable.svelte [hunk 10]
```diff
@@ -725,10 +725,11 @@ function handleSubmittedEdits(event: CustomEvent){
     let rowIndex: number = event.detail.rowIndex;
     let row = data[pageData[rowIndex][dataIdKey]];
     let rowInputs = (<HTMLInputElement>event.detail.domEvent.target).closest('tr')?.querySelectorAll(`[data-name=edit-input]`);
-    
-    if (!rowInputs) {
-        // ToDo: error should be easily visible to the end user
-        console.log('error');
+
+    const numRenderedColumns = instructs.length - 1 // minus one for checkbox column
+    if (rowInputs.length !== numRenderedColumns) {
```
Comment (mavcook) [inline app/src/lib/components/PowerTable.svelte]: This check means if a user wants to make a column non-editable, their "constant" component would still need to use the `data-name=edit-input` to not cause an error:
```svelte
<!-- my non-editable component -->
<script>export let instructKey;export let instructTitle;export let value;export let rowIndex;export let rowId;</script>
<div data-name=edit-input>
	<p>{value}</p>
</div>
```

File: app/src/routes/examples/example7/MyComponent.svelte (hunk 7 above)
Comment (mavcook) [reply app/src/routes/examples/example7/MyComponent.svelte]: I think an example of a non-editable component could be useful, perhaps for the `id` column.

File: app/src/lib/components/PowerTable.svelte (hunk 2 above)
Comment (muonw-public) [reply app/src/lib/components/PowerTable.svelte]: If in `src/lib/components/PowerTable.svelte` we have this

```js
export const SUBMISSION_EVENT: 'edits-submitted';
```

and later this

```js
<svelte:component this={instruct.edit?.as} {cell} {...instruct.edit?.props} on:edits-submitted={handleSubmittedEdits}/>
```

users have to always use the event name `edits-submitted` for any number of custom components they create. They can't define their own event names because it has to match what `<svelte:component>` inside PowerTable.svelte is listening to. Or maybe I'm missing your point. If that's the case, please let me know.

File: app/src/routes/examples/example7/MyComponent.svelte (hunk 7 above)
Comment (muonw-public) [reply app/src/routes/examples/example7/MyComponent.svelte]: > I think an example of a non-editable component could be useful, perhaps for the `id` column.

I didn't think about non-editable component. Yes, that would be cool.

File: app/src/lib/components/PowerTable.svelte (hunk 10 above)
Comment (muonw-public) [reply app/src/lib/components/PowerTable.svelte]: This needs to be tested. I haven't tried it but it seems like the cell with `<div data-name=edit-input>` would bypass the check but its corresponding value will get replaced by an empty string. Also having `data-name=edit-input` for a `div` is not ideal. Additionally, there might be cases that each cell would have more than one input. For example, an RGB representation of a color with inputs for Red, Green, and Blue numbers.

If you can find a way to make all edit components in a row report their final values once any submit button on that row in clicked, we won't have to deal with the complexity of each custom edit component and we won't have to resort to `querySelectorAll` to collect the values.

Comment (muonw-public) [reply app/src/lib/components/PowerTable.svelte]: > If you can find a way to make all edit components in a row report their final values once any submit button on that row in clicked, we won't have to deal with the complexity of each custom edit component and we won't have to resort to `querySelectorAll` to collect the values.

Actually, it's probably better (and also easier) if we keep track of all the "staged" (unsubmitted) changes in an Svelte store! So the edit components are responsible to write changes to this store and once an edit is submitted for row x, the staged changes of the row x will override the row data. This way even if we navigate back and forth between pages, we won't lose the staged changes. Also, we would be able to keep track of which cells have changed, so if user is closing the tab or exiting the edit mode, we can warn them about unsaved changes. This doesn't need to be a part of this PR though. For now an alert would suffice.

File: app/src/lib/components/PowerTable.svelte (hunk 2 above)
Comment (mavcook) [reply app/src/lib/components/PowerTable.svelte]: Sorry I didn't explain clearly, let me try again. I will use the new name ,`edit-submit-event`, from the most recent code (previously `edits-submitted`).

Say I am using PowerTable and have components A, B, C, and D I will use for editing cells. In each component `A.svelte`, `B.svelte`, and so on, I will have a function that is always implemented as such:
```typescript
function submitEdits(e: Event) {
    dispatch('edit-submit-event', {
        rowIndex: rowIndex,
        domEvent: e
    });
}
```

My next step would be to create a constant that I could use to replace the hardcoded string, say a`POWERTABLE_EDIT_SUBMIT_EVENT = 'edit-submit-event'` in `'MyConstants.svelte'` file. And then I can update the function in each A,B,C,D component to use the constant and get rid of the harcoded string:
```typescript
import POWERTABLE_EDIT_SUBMIT_EVENT from 'MyConstants.svelte';
function submitEdits(e: Event) {
    dispatch(POWERTABLE_EDIT_SUBMIT_EVENT, {
        rowIndex: rowIndex,
        domEvent: e
    });
}
```

Instead of having me, the client, define my own constant, it would be nice if PowerTable provided such a constant. So the import becomes `import POWERTABLE_EDIT_SUBMIT_EVENT from '@muonw/powertable';`.

File: app/src/lib/components/PowerTable.svelte (hunk 10 above)
Comment (mavcook) [reply app/src/lib/components/PowerTable.svelte]: Fortunately the value doesn't get set to an empty string, but I didn't do very robust testing. 

> Additionally, there might be cases that each cell would have more than one input. For example, an RGB representation of a color with inputs for Red, Green, and Blue numbers.

Interesting. That seems like a use case that would come up often. The current code for writing the data back to the `row` does not account for multiple inputs per component either, so I would need to update that too.

Let me think about this some more. I was actually looking into getting rid of `querySelectorAll` when I first started. From what I saw, there isn't an easy way to accomplish, but I didn't dig too much.

Comment (mavcook) [reply app/src/lib/components/PowerTable.svelte]: Oops, didn't see your most recent comment before responding yesterday.

Makes sense to do the Stores work in another PR. I actually did some work before seeing your comment, and got *something* working. Whether it is actually reasonable is another story. I'm meddling with powers I cannot possibly comprehend 😛 

In regards to getting multiple Inputs per edit-component working, I don't see a way to support that while relying on the `HTMLInputElement.value`. The client would need to provide one final value for us save, maybe via data attributes. I could work on that, or put effort towards the Stores approach?

File: app/src/lib/components/PowerTable.svelte (hunk 2 above)
Comment (muonw-public) [reply app/src/lib/components/PowerTable.svelte]: The first code is simpler but the second code is easier to maintain. Since there are other strings that users have to hard code (e.g. `rowIndex` key in the code above) and are similarly unlikely to change before a breaking change, hard coding this event name is fine.

File: app/src/lib/components/PowerTable.svelte (hunk 10 above)
Comment (muonw-public) [reply app/src/lib/components/PowerTable.svelte]: I think for now supporting only one input per cell is fine and would work for most users.

Comment (mavcook) [reply app/src/lib/components/PowerTable.svelte]: Ok how would you like to proceed? This PR should now be enough to fix the original issue, but the Stores approach sounds like it would be better, and would end up changing a lot of things. I could start work on new branch and close this PR?

Comment (muonw-public) [inline app/src/lib/components/PowerTable.svelte]: Typescript complains that `rowInputs` might be undefined.

This will result in the cell value to be replaced with an empty string (once an edit button is clicked):

```svelte
<div data-key={instructKey} data-name='edit-input'>
	{value}
</div>
```

and this will result in a new key (an empty string) to be added to the row:

```svelte
<div data-name='edit-input'>
	{value}
</div>
```

It would be better not to require `data-name='edit-input'` on an element like div. 

I haven't thought about the store idea enough to know exactly how it should be done, but if you end up implementing it for yourself and decide to share your code via a PR it would be appreciated.

**Review by muonw-public — CHANGES_REQUESTED** [review]

File: app/src/lib/components/PowerTable.svelte (hunk 10 above)
Comment (mavcook) [reply app/src/lib/components/PowerTable.svelte]: In regards to Typescript complaining, what do you think is the best way to handle the warning? The options I see are
- adding a `// @ts-ignore`, which is my initial preference
- return when `rowInputs` is `null`, but that should never really happen, and if it does, you got bigger worries

At the fear of wasting your time, and being too pedantic for a simple change, I've put my additional mulling behind a paywall as a deterrent.
<details><summary>$5 paywall</summary>

From what I understand, the warning is occurring because in the line
```typescript
let rowInputs = (<HTMLInputElement>event.detail.domEvent.target).closest('tr')?.querySelectorAll(`[data-name=edit-input]`);
```
 [`.closest('tr')` could return null](https://developer.mozilla.org/en-US/docs/Web/API/Element/closest#return_value) (and note that `.querySelectorAll()` always [returns an array](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelectorAll#return_value)). If it is the case that `rowInputs` is null, then the page somehow has no table rows (even a filtered table with no results still will have a `<tr>`). The check below can get rid of the warning, but probably isn't too useful to tell a user they don't have a table:
```typescript
    if (!rowInputs) {
        alert('No table rows found');
        return;
    }
```

Unless I am misunderstanding the error and behaviors, I think adding a `// @ts-ignore` here is most appropriate? Adding the null check code above fixes the issue, but would purely be for satisfying Typescript, and not the users. Let me know if I understand correctly, and which option you prefer me to add.
</details>

---

In regards to the non-editable div example, I am not sure how to get around the issues you mention. That is why I didn't end up adding such an example into the example7/ dir, as that would indicate official support. Apologies that I didn't communicate my thoughts there previously.

One easy way to solve the issues, but would remove the validation check you requested previously, is to remove the `if (rowInputs.length !== numRenderedColumns)` check. Then the non-editable example HTML could become
```
<div>
	<p>{value}</p>
</div>
```

Are ok with removing that check? Or maybe there is a more clever check to validate the user's edit-component has the necessary `data-name='edit-input`, but I haven't thought of one. An aside is below if you care to discuss in more depth why I think it is ok to remove the check.
<details><summary>aside</summary>

In my opinion, it shouldn't be up to the library at runtime to worry about how a client implemented a custom component. Since we don't have "compile time" ways to do these checks, or enforce an interface or template, the responsibility should fall on the library's documentation being good, and the developer reading said documentation to make sure they implemented things correctly. The only benefit of having a runtime error, that I can quickly think of, is that it makes bugs more triage-able.
</details>

Comment (mavcook) [reply app/src/lib/components/PowerTable.svelte]: For use of Svelte stores, I opened https://github.com/muonw/muonw-powertable/issues/34

Comment (muonw-public) [reply app/src/lib/components/PowerTable.svelte]: In this case `// @ts-ignore` makes sense, but is not ideal (it disables the warnings indiscriminately, you have to use it at least for two lines, and it wouldn't be clear why it's there).

Instead you can either change `.closest('tr')?` to `.closest('tr')!` to assert that there is going to be a `tr` tag, or you can add the null check to the same line you're checking the number of rendered columns:

```js
const numRenderedColumns = instructs.length - 1; // minus one for checkbox column
if (!rowInputs || rowInputs.length !== numRenderedColumns) {
    alert(`Only found ${rowInputs?.length} of ${numRenderedColumns} expected inputs. Make sure 'data-name=edit-input' attribute is on your component's input field`);
    console.error('Misconfigured edit component inputs');
    return;
}
```
<sub>Side note: I have also added `;` to the end of the first line.</sup>

You can remove the checks, but ideally we want to make sure to somehow handle those inputs that don't have the proper attributes (e.g. data-name). Otherwise user might edit the content and submit the changes without noticing that some of the changes were ignored. Developers should follow the documentations, but mistakes happen and we can at least try to make sure they won't go unnoticed whenever possible. 

```js
row[(<HTMLInputElement>rowInput)?.dataset?.key ?? ''] = (<HTMLInputElement>rowInput)?.value ?? '';
```

If `key` or `value` return null we don't want to substitute them with `''` and move on. So, this might be a good place to check and see if displaying an error is useful. Since `handleSubmittedEdits` is performed on a single row, we can afford to iterate through the cells with more flexibility. We can also use CSS to disable common form control elements of an edit block that aren't explicitly identified as `edit-input`, etc.

Comment (mavcook) [reply app/src/lib/components/PowerTable.svelte]: I liked your idea for checking `key` and `value`. Just pushed the changes for it.

File: app/src/routes/examples/example7/MyComponent.svelte (hunk 7 above)
Comment (mavcook) [reply app/src/routes/examples/example7/MyComponent.svelte]: Added in c294993

**Comment by mavcook** [comment]
Hi @muonw-public, all previous feedback has been addressed, so this PR should be good to land when you get a chance to do a final check. Let me know if there is anything else you would like to see, or if I can help in any way. Thanks!

**Comment by muonw-public** [comment]
It's unclear how to preserve the integrity of data with this change. In particular, if we have `props: {selectValues: ['Red', 'Blue', 'Green']}` in the instructs, but data already contains the value 'Yellow', how should that be handled during the edit.
Regardless, it makes sense to continue waiting for the release of Svelte 5 as it may call for a rewrite.
